src/util/logger.rs

```diff
@@ -0,0 +1,389 @@
+// Pruned copy of crate rust log, without global logger
+// https://github.com/rust-lang-nursery/log #7a60286
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::cmp;
+use std::fmt;
+use std::str::FromStr;
+use std::sync::Arc;
+
+use std::ascii::AsciiExt;
+
+static LOG_LEVEL_NAMES: [&'static str; 6] = ["OFF", "ERROR", "WARN", "INFO", "DEBUG", "TRACE"];
+
+static LEVEL_PARSE_ERROR: &'static str =
+		"attempted to convert a string that doesn't match an existing log level";
+
+/// An enum representing the available verbosity levels of the logger.
+#[derive(Copy, Eq, Debug, Hash)]
+pub enum Level {
+	///Designates logger being silent
+	Off,
+	/// Designates very serious errors
+	Error = 1,
+	/// Designates hazardous situations
+	Warn,
+	/// Designates useful information
+	Info,
+	/// Designates lower priority information
+	Debug,
+	/// Designates very low priority, often extremely verbose, information
+	Trace,
+}
+
+impl Clone for Level {
+	#[inline]
+	fn clone(&self) -> Level {
+		*self
+	}
+}
+
+impl PartialEq for Level {
+	#[inline]
+	fn eq(&self, other: &Level) -> bool {
+		*self as usize == *other as usize
+	}
+}
+
+impl PartialOrd for Level {
+	#[inline]
+	fn partial_cmp(&self, other: &Level) -> Option<cmp::Ordering> {
+		Some(self.cmp(other))
+	}
+
+	#[inline]
+	fn lt(&self, other: &Level) -> bool {
+		(*self as usize) < *other as usize
+	}
+
+	#[inline]
+	fn le(&self, other: &Level) -> bool {
+		*self as usize <= *other as usize
+	}
+
+	#[inline]
+	fn gt(&self, other: &Level) -> bool {
+		*self as usize > *other as usize
+	}
+
+	#[inline]
+	fn ge(&self, other: &Level) -> bool {
+		*self as usize >= *other as usize
+	}
+}
+
+impl Ord for Level {
+	#[inline]
+	fn cmp(&self, other: &Level) -> cmp::Ordering {
+		(*self as usize).cmp(&(*other as usize))
+	}
+}
+
+fn ok_or<T, E>(t: Option<T>, e: E) -> Result<T, E> {
+	match t {
+		Some(t) => Ok(t),
+		None => Err(e),
+	}
+}
+
+impl FromStr for Level {
```

Comment:
Can we just not implement FromStr for Level? I'm not sure who would ever really need that, and it adds a whole Err type and associated impls.